File: 链表/86.分隔链表.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        bigger_equal = ListNode(0, None)

        less_p = dummy
        bigger_equal_p = bigger_equal

        cur = head
        while cur:
            if cur.val < x:
                less_p.next = cur
                less_p = less_p.next
            else:
                bigger_equal_p.next = cur
                bigger_equal_p = bigger_equal_p.next
            cur = cur.next
        # 连接
        less_p.next = bigger_equal.next
        bigger_equal_p.next = None  # 从最后的节点断开
        return dummy.next

File: 链表/86.分隔链表.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} x
 * @return {ListNode}
 */
 //! 当移动数组中的节点时断开与原数组的连接
 var partition = function (head, x) {
    const leftDummy = new ListNode();
    const rightDummy = new ListNode();
    let left = leftDummy, right = rightDummy, cur = head;
  
    while (cur) {
      // <x在左边, >=x在右边
      if (cur.val < x) {
        left.next = cur;
        left = left.next;
      } else {
        right.next = cur;
        right = right.next;
      }
      const nextNode = cur.next; // 保存下一个节点
      cur.next = null; // 断开当前节点的原链表连接，防止形成环
      cur = nextNode; // 移动到下一个节点
    }
    left.next = rightDummy.next;
    return leftDummy.next;
  };
  